### PR TITLE
perf: DH-23690: Plan SortedRanges bulk inserts and removals before moving entries

### PR DESCRIPTION
**Pull request by cpwright**
SortedRanges.insert(RowSet) and remove(RowSet) merged the whole set with
the argument entry by entry through range iterators and then repacked the
result, so adding two scattered keys to a 2000-entry set cost 40-68 times
what inserting them one at a time did. Incremental naturalJoin accumulates
each slot's left row keys per cycle and applies them with one such bulk
call, which is where the DH-23190 regression on sparse updates came from.

Both operations now run in two passes over the argument's ranges. The
first plans every edit: a binary search per range, ranges that touch
existing ones coalescing into one, cut ranges leaving zero to two
remainders, consecutive ranges cutting the same range carving one edit.
The second moves each untouched stretch of entries exactly once, by the
net entry delta before it, backward in place when no edit shrinks, forward
in place when none grows, and into a new set of the same type otherwise or
when the set is shared or too small. One or two ranges, or a handful on a
tiny set, still take the individual add/remove path, and an argument with
more than an eighth as many entries as the set still takes the merge; the
benchmarks show each is cheaper in its band. When the result outgrows the
set's type the existing union/intersect path repacks it as before.

RowSetSmallInsertBench and RowSetSmallRemoveBench compare the bulk call
against forAllRowKeys on the same keys. Bulk insert is within noise of
key-by-key at one or two keys and 2-5x faster from five keys up (2000
entries, 20 keys: 0.48 vs 0.97 us; 6000 entries, 100 keys: 4.0 vs 12.3
us); bulk remove likewise (6000 entries, 100 keys: 2.0 vs 10.4 us). With
this in place accumulation beats direct per-key insertion on every
naturalJoin key-churn shape measured, including 2000-key slots at five
inserts per slot where it previously lost by 84%.

SortedRangesBulkInsertTest checks insert and remove against key-by-key
across short, int and long packings, shared copies, adjacency, overlap,
bridging, splits, and ranges reaching Long.MAX_VALUE.

### CODE REVIEW

File: engine/benchmark/src/main/java/io/deephaven/benchmark/engine/util/RowSetSmallInsertBench.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    @Setup(Level.Invocation)
+    public void setupInvocation() {
+        target = buildTarget();
+    }
+
+    @TearDown(Level.Invocation)
+    public void tearDownInvocation() {
+        target.close();
+    }
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/benchmark/src/main/java/io/deephaven/benchmark/engine/util/RowSetSmallInsertBench.java:121]: Added a trial teardown closing base and added.

File: engine/benchmark/src/main/java/io/deephaven/benchmark/engine/util/RowSetSmallRemoveBench.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    @Setup(Level.Invocation)
+    public void setupInvocation() {
+        target = buildTarget();
+    }
+
+    @TearDown(Level.Invocation)
+    public void tearDownInvocation() {
+        target.close();
+    }
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/benchmark/src/main/java/io/deephaven/benchmark/engine/util/RowSetSmallRemoveBench.java:136]: Added a trial teardown closing removed.

File: engine/rowset/src/test/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRangesBulkInsertTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                    }
+
+                    // A shared copy: the insert must leave the original untouched.
+                    try (final WritableRowSet shared = target.copy()) {
+                        shared.insert(added);
+                        check(trial, expected, shared);
+                        assertEquals("trial " + trial + " original changed", targetInner.ixCardinality(),
+                                target.size());
+                        assertTrue("trial " + trial + " original changed", target.subsetOf(expected));
+                    }
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/src/test/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRangesBulkInsertTest.java]: The shared-copy checks now compare the original against an independent key-by-key snapshot taken before the edit.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java [hunk 4]
```diff
@@ -1809,6 +1810,581 @@ public final SortedRanges mergeAppend(final SortedRanges other, final boolean wr
...
+                if (next < 0) {
+                    end = -next;
+                    ++i;
+                }
+            }
+            result = result.addRangeInternal(start, end, check);
+            if (result == null) {
+                return null;
+            }
+            check = false;
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java]: Good catch. The check now stays on until an insert returns a set other than this one, which is the private copy; containedFirstRangeKeepsSharedCopyIsolated covers the contained-first-range sequence.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java [hunk 5]
```diff
@@ -1809,6 +1810,598 @@ public final SortedRanges mergeAppend(final SortedRanges other, final boolean wr
...
+            } else {
+                final long data = unpackedGet(p);
+                if (data < 0) {
+                    // The range ending here started before s - 1 and reaches at least s - 1: it touches [s, e].
+                    final long rangeStart = unpackedGet(p - 1);
+                    g0 = p - 1;
+                    g1 = p + 1;
+                    gFirst = Math.min(rangeStart, s);
+                    gLast = Math.max(-data, e);
+                    gOldCardinality = -data - rangeStart + 1;
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java]: The absorbTouching walk runs after both branches of that if/else (it follows the else block), so the range at 51 is absorbed and the result is [48,51]. Added bridgingInsertCoalescesThroughToTheNextRange with exactly this shape, through the individual and planned paths, private and shared.

File: engine/rowset/docs/sortedranges-bulk-edits.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+| **Benchmarks** | `RowSetSmallInsertBench`, `RowSetSmallRemoveBench` (`jmhRunRowSetSmallInsert`, `jmhRunRowSetSmallRemove`) |
+| **Tests** | `SortedRangesBulkInsertTest` (randomized against key-by-key, plus the named corner cases below) |
+
+## The representation the strategies edit
+
+A `SortedRanges` is one sorted array of *entries*. A non-negative entry is a key that starts a range or stands alone
+as a single; a negative entry is the (negated) end of the range whose start precedes it. So `{3, 10..12, 20}` is stored
+as `3, 10, -12, 20`: four entries for three ranges. Ranges are never adjacent or overlapping; `{10} {11..12}` is not a
+valid state, it is `10..12`. The array is packed as shorts, ints or longs (`SortedRangesShort/Int/Long`) relative to an
+offset, and each packing has a capacity beyond which the set must become an `RspBitmap`.
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/docs/sortedranges-bulk-edits.md]: Reworded: shorts and ints are offset-relative, longs are stored raw; a capacity overflow is repacked into another packing when one can hold it, and only a result no packing holds becomes an RspBitmap. The tree, append and merge sections say the same.

File: engine/rowset/docs/sortedranges-bulk-edits.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+search and one `System.arraycopy` to open or close a gap, moving about half the entries. That block move is cheap,
+about 0.02-0.05 ns per entry moved, which is why two such moves beat one planned pass with its fixed setup.
+
+Copy-on-write: `removeRange` handles a shared receiver itself. `addRangeInternal(..., writeCheck)` does too, but
+returns the receiver *unchanged and still shared* when the range is already contained, so the loop keeps `writeCheck`
+on until a call returns a different object, which is the private copy. Clearing it after the first call let a later
+range write into a shared set; `containedFirstRangeKeepsSharedCopyIsolated` covers the sequence.
+
+A null from either loop means a range outgrew the set's type. The loop may already have applied earlier ranges in
+place; the caller falls through to the merge, for which re-inserting or re-removing them is idempotent.
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/docs/sortedranges-bulk-edits.md]: Reworded as a result-capacity failure that covers both loops: too many entries after inserting, or after splitting on removal.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java [hunk 8]
```diff
@@ -4089,6 +4748,17 @@ public final OrderedLongSet ixRemove(final OrderedLongSet removed) {
     }
 
     public final OrderedLongSet remove(final OrderedLongSet removed) {
+        if (removed instanceof SortedRanges) {
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java:4781]: Good catch. remove(OrderedLongSet) now returns EMPTY when the argument is the receiver, before choosing a strategy; editingASetWithItself covers insert and remove of a set with itself on the individual and planned paths, private and shared.

File: engine/rowset/src/test/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRangesBulkInsertTest.java [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+                builder.appendKey(key);
+            }
+            try (final WritableRowSet set = builder.build()) {
+                try (final WritableRowSet inserted = snapshot(set)) {
+                    inserted.insert(inserted);
+                    assertTrue("keys=" + keys, set.equals(inserted));
+                }
+                try (final WritableRowSet shared = set.copy();
+                        final WritableRowSet snapshotBefore = snapshot(set)) {
+                    shared.insert(shared);
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/src/test/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRangesBulkInsertTest.java]: The test now aliases through two wrappers sharing one inner set (shared.insert(set) / shared.remove(set)), which reaches the SortedRanges strategies and the identity guard while keeping the isolation assertion on the other wrapper.

File: engine/rowset/src/test/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRangesBulkInsertTest.java [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+                shared.remove(removed);
+                check(0, expected, shared);
+                assertTrue(snapshotBefore.equals(target));
+            }
+        }
+    }
+
+    /**
+     * Removing a set from itself must empty it, and inserting a set into itself must change nothing, on every strategy
+     * and whether or not the set is shared.
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/src/test/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRangesBulkInsertTest.java]: Renamed and rescoped: the Javadoc now says which paths the same-sized alias selects (individual at two ranges, merge above; the identity guard for removal).

File: engine/rowset/docs/sortedranges-bulk-edits.md [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+- **Removing from a set an earlier removal emptied.** `removeRange` on an empty set fails; the individual loop stops
+  when the result is empty.
+- **The argument is a normalized set.** Its ranges never touch, which is what makes the carve invariant hold and what
+  made a proposed test case (`{10} {11..12}`) impossible rather than uncovered.
+- **Small arguments are not the only arguments.** The first cut of the planned strategy had no upper bound and was
+  up to 2.3x slower than the merge for arguments near the set's own size; `planEdits` is the result. Any change to
+  these strategies should be measured across the whole `k` axis of both benchmarks, not only the small end.
+- **The argument may be the receiver.** `rowSet.remove(rowSet)` hands the individual remove loop the set it is
+  reading its ranges from; the merge handled it, the loop would cut and re-read. `remove` returns the empty set for
+  its own receiver before choosing a strategy. Inserting a set into itself is a no-op on every path.
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/docs/sortedranges-bulk-edits.md]: Reworded: the wrapper handles rowSet.remove(rowSet) itself; the SortedRanges-level case is two wrappers sharing an inner set.

**Comment by cpwright** [comment]
## Benchmark results

`RowSetSmallInsertBench` and `RowSetSmallRemoveBench`, microseconds per operation, on a 24-core Linux box with nothing else running. Targets are scattered single keys: a `SortedRangesInt` up to 6000 entries, an `RspBitmap` at 20000. Columns:

- **before**: `main`'s `SortedRanges`, where every bulk call merges the two sets (the two source files swapped in on this branch, so the harness is identical).
- **after**: this PR's head.
- **forAll**: the same keys applied one at a time through `forAllRowKeys`, the comparison baseline.

The last two columns are speedups of the PR over each baseline, so larger is better. A speedup is **bold** when it is at least 1.33x (25% less time) and outside the measurement noise. The noise floor is 0.1 µs absolute on any cell, because the sub-microsecond rows moved by up to that between identical runs of one build, and 10% relative otherwise, widened to 20% on the 20000-entry rows, whose identical-code cells moved that much between runs. Every cell that reads as a slowdown falls inside one of those floors.

Rows with more keys than an eighth of the target (500 keys and up on 2000 entries, 2000 and up on 6000) take the merge by design, so before and after are the same code there. The 20000-entry rows are the `RspBitmap` path, untouched by this change.

### Insert

| target entries | keys | before (merge) µs | after µs | forAll µs | before / after | forAll / after |
|---:|---:|---:|---:|---:|---:|---:|
| 20 | 1 | 0.023 | 0.023 | 0.023 | 1.00x | 1.00x |
| 20 | 2 | 0.080 | 0.041 | 0.040 | 1.95x | 0.98x |
| 20 | 5 | 0.096 | 0.099 | 0.078 | 0.97x | 0.79x |
| 20 | 20 | 0.142 | 0.137 | 0.265 | 1.04x | **1.93x** |
| 20 | 100 | 0.543 | 0.430 | 1.411 | 1.26x | **3.28x** |
| 200 | 1 | 0.027 | 0.027 | 0.028 | 1.00x | 1.04x |
| 200 | 2 | 0.399 | 0.068 | 0.066 | **5.87x** | 0.97x |
| 200 | 5 | 0.425 | 0.165 | 0.128 | **2.58x** | 0.78x |
| 200 | 20 | 0.458 | 0.459 | 0.373 | 1.00x | 0.81x |
| 200 | 100 | 0.821 | 0.822 | 1.686 | 1.00x | **2.05x** |
| 200 | 500 | 2.110 | 2.139 | 11.105 | 0.99x | **5.19x** |
| 200 | 2000 | 6.965 | 6.569 | 61.673 | 1.06x | **9.39x** |
| 200 | 6000 | 17.327 | 17.216 | 245.174 | 1.01x | **14.2x** |
| 2000 | 1 | 0.067 | 0.066 | 0.075 | 1.02x | 1.14x |
| 2000 | 2 | 4.233 | 0.111 | 0.104 | **38.1x** | 0.94x |
| 2000 | 5 | 4.531 | 0.206 | 0.266 | **22.0x** | 1.29x |
| 2000 | 20 | 4.555 | 0.530 | 0.969 | **8.59x** | **1.83x** |
| 2000 | 100 | 5.421 | 2.751 | 5.071 | **1.97x** | **1.84x** |
| 2000 | 500 | 6.447 | 6.571 | 25.463 | 0.98x | **3.88x** |
| 2000 | 2000 | 11.041 | 10.821 | 221.896 | 1.02x | **20.5x** |
| 2000 | 6000 | 23.187 | 23.263 | 552.990 | 1.00x | **23.8x** |
| 6000 | 1 | 0.089 | 0.091 | 0.129 | 0.98x | 1.42x |
| 6000 | 2 | 13.006 | 0.164 | 0.198 | **79.3x** | 1.21x |
| 6000 | 5 | 12.929 | 0.333 | 0.611 | **38.8x** | **1.83x** |
| 6000 | 20 | 15.151 | 2.215 | 4.196 | **6.84x** | **1.89x** |
| 6000 | 100 | 15.157 | 4.167 | 12.239 | **3.64x** | **2.94x** |
| 6000 | 500 | 17.790 | 12.720 | 63.188 | **1.40x** | **4.97x** |
| 6000 | 2000 | 23.223 | 23.220 | 283.364 | 1.00x | **12.2x** |
| 6000 | 6000 | 372.724 | 378.971 | 4795.968 | 0.98x | **12.7x** |
| 20000 | 1 | 27.154 | 27.645 | 28.537 | 0.98x | 1.03x |
| 20000 | 2 | 33.694 | 34.478 | 29.333 | 0.98x | 0.85x |
| 20000 | 5 | 29.963 | 30.374 | 42.194 | 0.99x | **1.39x** |
| 20000 | 20 | 41.263 | 43.698 | 287.089 | 0.94x | **6.57x** |
| 20000 | 100 | 49.970 | 50.541 | 1483.166 | 0.99x | **29.3x** |

### Remove

| target entries | keys | before (merge) µs | after µs | forAll µs | before / after | forAll / after |
|---:|---:|---:|---:|---:|---:|---:|
| 20 | 1 | 0.022 | 0.022 | 0.020 | 1.00x | 0.91x |
| 20 | 2 | 0.080 | 0.031 | 0.032 | 2.58x | 1.03x |
| 20 | 5 | 0.087 | 0.062 | 0.060 | 1.40x | 0.97x |
| 20 | 20 | 0.152 | 0.154 | 0.147 | 0.99x | 0.95x |
| 200 | 1 | 0.026 | 0.026 | 0.024 | 1.00x | 0.92x |
| 200 | 2 | 0.469 | 0.040 | 0.039 | **11.7x** | 0.97x |
| 200 | 5 | 0.445 | 0.100 | 0.075 | **4.45x** | 0.75x |
| 200 | 20 | 0.467 | 0.288 | 0.295 | **1.62x** | 1.02x |
| 200 | 100 | 0.783 | 0.783 | 1.329 | 1.00x | **1.70x** |
| 2000 | 1 | 0.081 | 0.081 | 0.080 | 1.00x | 0.99x |
| 2000 | 2 | 5.017 | 0.154 | 0.153 | **32.6x** | 0.99x |
| 2000 | 5 | 5.084 | 0.181 | 0.319 | **28.1x** | **1.76x** |
| 2000 | 20 | 4.360 | 0.424 | 0.866 | **10.3x** | **2.04x** |
| 2000 | 100 | 4.701 | 1.762 | 4.101 | **2.67x** | **2.33x** |
| 2000 | 500 | 5.907 | 5.966 | 20.925 | 0.99x | **3.51x** |
| 2000 | 2000 | 11.172 | 11.128 | 71.781 | 1.00x | **6.45x** |
| 6000 | 1 | 0.203 | 0.199 | 0.241 | 1.02x | 1.21x |
| 6000 | 2 | 14.669 | 0.295 | 0.329 | **49.7x** | 1.12x |
| 6000 | 5 | 14.831 | 0.334 | 0.647 | **44.4x** | **1.94x** |
| 6000 | 20 | 15.455 | 0.547 | 2.053 | **28.3x** | **3.75x** |
| 6000 | 100 | 13.584 | 2.037 | 10.366 | **6.67x** | **5.09x** |
| 6000 | 500 | 14.811 | 8.618 | 57.919 | **1.72x** | **6.72x** |
| 6000 | 2000 | 25.542 | 25.742 | 243.714 | 0.99x | **9.47x** |
| 6000 | 6000 | 32.664 | 33.602 | 611.795 | 0.97x | **18.2x** |
| 20000 | 1 | 6.909 | 7.115 | 8.207 | 0.97x | 1.15x |
| 20000 | 2 | 27.727 | 28.081 | 44.745 | 0.99x | **1.59x** |
| 20000 | 5 | 35.043 | 34.096 | 119.051 | 1.03x | **3.49x** |
| 20000 | 20 | 38.227 | 38.263 | 435.815 | 1.00x | **11.4x** |
| 20000 | 100 | 44.501 | 45.586 | 1732.120 | 0.98x | **38.0x** |

The 20 by 100 remove cell is absent because the benchmark now rejects a removal of more keys than the target holds rather than measuring fewer under that label.

File: engine/rowset/docs/sortedranges-bulk-edits.md [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+## The representation the strategies edit
+
+A `SortedRanges` is one sorted array of *entries*. A non-negative entry is a key that starts a range or stands alone
+as a single; a negative entry is the (negated) end of the range whose start precedes it. So `{3, 10..12, 20}` is stored
+as `3, 10, -12, 20`: four entries for three ranges. Ranges are never adjacent or overlapping; `{10} {11..12}` is not a
+valid state, it is `10..12`. The array is packed as shorts or ints relative to an offset (`SortedRangesShort`,
+`SortedRangesInt`, via `SortedRangesPacked`) or as raw longs with no offset (`SortedRangesLong`). Each packing has a
+capacity. The merge-based paths choose the result's packing afresh, by span, density and capacity, as described under
+the merge below; the append path does not, and converts straight to an `RspBitmap` when its result does not fit our
+own packing. A result no packing can hold, or one the merge judges dense, becomes an `RspBitmap`.
```
[the remark this answers is not in the document]
Comment (cpwright) [reply engine/rowset/docs/sortedranges-bulk-edits.md]: Reworded both passages: an append keeps our packing, repacks only into a narrower one when capacity runs out, and converts to a bitmap only when the result is outside the packing's span or no packing has room.

File: engine/rowset/docs/sortedranges-bulk-edits.md [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+**`editIndividually(other)`**: true for one or two ranges at any size; beyond that, with `r` the number of ranges in
+`other` and `n` our entry count, true when `(r - 2)^2 * n <= 400`, where 400 is the default of
+`SortedRanges.individualEditThreshold`, which admits a handful of ranges on a tiny set. Ranges are counted from their
+start entries, stopping as soon as the inequality fails, so the count is cheap. This is where the planned
+strategy's fixed cost of a few tens of nanoseconds does not pay: measured, individual edits win for one or two ranges
+everywhere and for up to about six ranges on a 20-entry set.
+
+**`planEdits(other)`**: `other.count * 8 <= count`, on entries, where 8 is the default of
+`SortedRanges.plannedEditMaxSizeRatio`. A planned edit costs a binary search and a small block
+move, about 20 ns; the merge costs about 2 ns per entry of either set. Measured, planning wins while the argument has
```
Comment (jcferretti) [inline engine/rowset/docs/sortedranges-bulk-edits.md:97]: A general comment about mentioning measurements:
* A number alone has no way to being anchored.  We can anchor it by mentioning explicitly in some part of the doc where the measurements were made (machine type, CPU, etc).  Another way to anchor a number is to compare it with another number (eg, when  you do it this way, X, when you do it that way, Y).  The document is already doing that I think.  So I believe it would be good just to mention somewhere where the numbers are coming from, because in 3 years from now computers will be different.

Comment (cpwright) [reply engine/rowset/docs/sortedranges-bulk-edits.md:97]: Machine details included now.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java [hunk 14]
```diff
@@ -1809,6 +1810,654 @@ public final SortedRanges mergeAppend(final SortedRanges other, final boolean wr
...
+        }
+
+        int pieceEnd(final int edit) {
+            return edit + 1 < size ? pieceStart[edit + 1] : pieces;
+        }
+
+        /** Entries the edit's pieces occupy. */
+        int newLength(final int edit) {
+            int length = 0;
+            for (int pi = pieceStart[edit]; pi < pieceEnd(edit); ++pi) {
```
Comment (jcferretti) [inline engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java]: edit is final, and pieceEnd(edit) has a condition evaluation that should result in a constant too.  The compiler will likely realize that and avoid recalculating on every for loop, but reading the code as is makes me have to think about it, so, not necessarily to help the compiler, but to help me, can we move `pieceEnd(edit)` to a constant calculated before the loop?

Since the same loop is in two pieces I thought about suggesting something like a `forEachPiece` private method receiving a lambda, but modifying a primitive accumulator gets annoying (who wants additional MutableX thingies, that makes the thing more complicated than the two for loops).  Sigh java.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java [hunk 15]
```diff
@@ -1809,6 +1810,654 @@ public final SortedRanges mergeAppend(final SortedRanges other, final boolean wr
...
+
+    /**
+     * Walk our ranges from entry {@code ii} while they start at or before {@code bound + 1}, all of which touch a range
+     * ending at {@code bound}. Sets the plan's {@code absorbedCardinality} to the keys those ranges held and
+     * {@code absorbedLastEnd} to the last key any of them or {@code bound} reaches. The plan is thread-local scratch;
+     * this set may be shared between threads and is never written here.
+     *
+     * @return the entry after the last range absorbed
+     */
+    private int absorbTouching(final EditPlan plan, int ii, final long bound) {
```
Comment (jcferretti) [inline engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java:2155]: bad touch? (1999 Bloodhound Gang's song)

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java [hunk 16]
```diff
@@ -1809,6 +1810,654 @@ public final SortedRanges mergeAppend(final SortedRanges other, final boolean wr
...
+            if (plan.first[pi] != plan.last[pi]) {
+                sr.unpackedSet(pos++, -plan.last[pi]);
+            }
+        }
+    }
+
+    /**
+     * Apply a plan whose edits never shrink, in place: from the back, so every stretch moves before it is overwritten.
+     */
+    private void applyPlanBackward(final EditPlan plan) {
```
Comment (jcferretti) [inline engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java:2409]: This function name made me think of a report I once had.

File: engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java [hunk 17]
```diff
@@ -4089,6 +4764,22 @@ public final OrderedLongSet ixRemove(final OrderedLongSet removed) {
...
 
     public final OrderedLongSet remove(final OrderedLongSet removed) {
+        if (removed == this) {
+            // Removing a set from itself empties it; the individual path below would otherwise be cutting the very
+            // set it is reading its ranges from.
+            return OrderedLongSet.EMPTY;
+        }
+        if (removed instanceof SortedRanges) {
+            final SortedRanges removedSar = (SortedRanges) removed;
+            final SortedRanges ans = editIndividually(removedSar) ? removeRangesIndividually(removedSar)
```
Comment (jcferretti) [inline engine/rowset/src/main/java/io/deephaven/engine/rowset/impl/sortedranges/SortedRanges.java]: I can never read nested ternary operators without parenthesis, could you add some?

**Review by jcferretti — DISMISSED** [review]

**Review by jcferretti — APPROVED** [review]
